src/java.base/share/classes/java/lang/foreign/SegmentAllocator.java

```diff
@@ -257,89 +257,171 @@ default MemorySegment allocateFrom(AddressLayout layout, MemorySegment value) {
         return segment;
     }
 
+    /**
+     * Allocates a memory segment with the given layout and initializes it with the bytes in the provided
+     * source memory segment.
+     * @implSpec the default implementation for this method calls {@code this.allocate(elementLayout, elementCount)}.
+     * @param elementLayout the element layout of the allocated array.
+     * @param source the source segment.
+     * @param sourceElementLayout the element layout of the source segment.
+     * @param sourceOffset the starting offset, in bytes, of the source segment.
+     * @param elementCount the number of elements in the source segment to be copied.
+     * @return a segment for the newly allocated memory block.
+     * @throws IllegalArgumentException if {@code elementLayout.byteSize() != sourceElementLayout.byteSize()}.
+     * @throws IllegalArgumentException if the source segment/offset are <a href="MemorySegment.html#segment-alignment">incompatible with the alignment constraint</a>
+     * in the source element layout.
+     * @throws IllegalArgumentException if {@code elementLayout.byteAlignment() > elementLayout.byteSize()}.
+     * @throws IllegalArgumentException if {@code sourceElementLayout.byteAlignment() > sourceElementLayout.byteSize()}.
```

Comment:
Note: I want to rewrite and split similar assertions in the javadoc of some methods in MemorySegment too - but I'll do that as a separate patch (as it's going to be a bit noisy).